evm/src/executor/backend/mod.rs

```diff
@@ -661,6 +703,62 @@ impl Backend {
         }
         Ok(())
     }
+
+    /// Returns the block numbers required for replaying a transaction
+    fn get_block_numbers_for_transaction(
+        &self,
+        id: LocalForkId,
+        transaction: H256,
+    ) -> eyre::Result<(U64, U64)> {
+        let fork = self.inner.get_fork_by_id(id)?;
+        let tx = fork.db.db.get_transaction(transaction)?;
+
+        // get the block number we need to fork
+        if let Some(tx_block) = tx.block_number {
+            // we need to subtract 1 here because we want the state before the transaction
+            // was mined
+            let fork_block = tx_block - 1;
+            Ok((fork_block, tx_block))
+        } else {
+            let block = fork
+                .db
+                .db
+                .get_full_block(BlockNumber::Latest)?
+                .number
+                .ok_or_else(|| DatabaseError::BlockNotFound(BlockNumber::Latest.into()))?;
+            Ok((block, block))
+        }
+    }
+
+    /// Replays all the transactions at the forks current block that were mined before the `tx`
+    ///
+    /// Returns the _unmined_ transaction that corresponds to the given `tx_hash`
+    pub fn replay_until(
+        &mut self,
+        id: LocalForkId,
+        mut env: Env,
+        tx_hash: H256,
+    ) -> eyre::Result<Option<Transaction>> {
+        let fork = self.inner.get_fork_by_id_mut(id)?;
+        let full_block =
+            fork.db.db.get_full_block(BlockNumber::Number(env.block.number.as_u64().into()))?;
+
+        for tx in full_block.transactions.into_iter() {
+            if tx.hash().eq(&tx_hash) {
+                // found the target transaction
+                return Ok(Some(tx))
+            }
```

Comment:
can we cache this somehow to disk? so that forking a txhash does not replay all txs each time, and can maybe save the pre-tx hash state (or including it)?

Reply:
would require adding another mapping to how forks are tracked rn, basically (endpoint, block, tx index).

would be possible but requires a bit of refactoring effort, I'd leave this for a follow up for perf